encoredev/encore
# Enable Tests to be written against PubSub topics

This commit adds a new implementation of a pubsub topic; the test implementation. This Implementation is automatically aware of the test currently running, and isolates it's behaviour to that test.

This means you can run two separate tests in parallel, both of which interact with the same `pubsub.Topic`, yet they will only observe the side-effects of their own published messages.

To test a pubsub topic;

```go
import (
    "context"
    "testing"

    "github.com/stretchr/testify/require"

    "encore.dev/pubsub"
    "encore.dev/test"
)

type Msg struct { Str string }

var myTopic = pubsub.NewTopic[Msg]("my-topic")

func TestOne(t *testing.T) {
    t.Parallel()

    err := myTopic.Publish(context.Background(), Msg{"Hello World"})
    require.NoError(t, err)

    // even though these two tests run in parallel, the message from TestTwo can not appear here.
    messages := test.Topic(myTopic).PublishedMessages()
    require.Len(t, messages, 1)
    require.Equal(t, "Hello World", messages[0].Str)
}

func TestTwo(t *testing.T) {
    t.Parallel()

    err := myTopic.Publish(context.Background(), Msg{"Goodbye World"})
    require.NoError(t, err)
}
```

To achieve this we've added four new linked functions between the Go runtime and the Encore runtime in encoredev/go#8 . These functions inform the Encore runtime when tests are started, paused, resumed and finished. The Encore runtime then tags the running go routine with the test information. (See the new `test_interface.go` file)

This then allows the new `pubsub/test.Topic` implementation to access the current test, and assign it a specific implementation instance of a topic.

The `test.Topic()` function then uses a private API to extract that specific instance for a test and casts it into a public interface, allowing user code to perform assertions upon the messages published on the topic.

Future expansions to this implementation will allow the user to forcible test failure scenarios.
